Forward gRPC context when using @blocking

gaetancollaud opened the pull request:
closes #14665 

Since #14687 seems rather complicated for nothing I created this PR.

It simply captures the gRPC context and inject it in the blocking handler.

gsmet reviewed (changes requested):
Could you squash the 2 commits?

Also would it be possible to add a test?

gaetancollaud commented:
> Could you squash the 2 commits?
> 
> Also would it be possible to add a test?

I squashed manually my commits. Just out of curiosity, why don't you enable the "Squash mergin" option of github ? It would make thinks easier for developer, no ?

I added some tests, but I had to mock a lot to just test this interceptor. Is it ok for you like this ?

gsmet commented:
> I squashed manually my commits. Just out of curiosity, why don't you enable the "squash merging" option of github ? It would make thinks easier for developer, no ?

It doesn't work with merge commits.

gsmet commented:
@cescoffier I let you have a look at the test? I wonder if we should backport this to 1.11 given it's more or less a bugfix. WDYT?

cescoffier reviewed (approved):
You would have preferred a real test, but it's good for now.

gsmet commented on extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/BlockingServerInterceptor.java:
```diff
@@ -96,11 +97,17 @@ synchronized void setDelegate(ServerCall.Listener<ReqT> delegate) {
                         consumer.accept(delegate);
                         f.complete();
+
+                        grpcContext.detach(previous);
```
@cescoffier maybe it's not mandatory in this case but we usually restore previous contexts in a finally block. It's not needed here?

gaetancollaud replied:
I think it's best to keep it. In my opinion this part of the code is responsible to wrap the context. And it should not rely on any other code that will eventually drop the context. This could have sever security implication. I use the context to store the JWT token. If the context is not cleared correctly and somehow the jwt end up in another request it's a security leak.

gsmet replied:
Maybe I wasn't clear.

My question was: shouldn't this code look like:
```java
final Context previous = Context.current();
grpcContext.attach();
try {
    consumer.accept(delegate);
    f.complete();
} finally {
    grpcContext.detach(previous);
}
```

gsmet replied:
(you might want to put the `attach` in the `try` too, no idea what it does :))

gaetancollaud replied:
Hoo my bad, then I agree, we should do the detach in the finally.

Not sure about the attach. But in the javadoc example it's outside the try: https://grpc.github.io/grpc-java/javadoc/io/grpc/Context.html#attach--

gaetancollaud replied:
@gsmet done

gaetancollaud committed:
[gRPC] Forward Grpc context when using @blocking

gsmet reviewed (approved):
Thanks!

gsmet commented:
And merged, thanks!
